File: generateMultiWaypoint.m
function result = generateMultiWaypoint(robot, waypoints_xyz, orientations, toolSpeeds)
%% generateMultiWaypoint  Multi-waypoint cubic Hermite trajectory.
%%
%%   result = generateMultiWaypoint(robot, waypoints_xyz, orientations, toolSpeeds)
%%
%%   Inputs
%%     robot         — loadrobot result with Gravity set
%%     waypoints_xyz — (nWP x 3) matrix of [x y z] targets
%%     orientations  — (nWP x 1) cell of strings: 'Down (-Z)','Up (+Z)',
%%                     'Horizontal (+X)','Tilted 45°'
%%     toolSpeeds    — scalar OR (nWP x 1) vector of speeds in m/s
%%
%%   Output result struct: same fields as generateTrajectory

ee     = 'EndEffector_Link';
tauMax = [187;187;187;52;52;52;52];
dt     = 0.001;
nWP    = size(waypoints_xyz, 1);

if nargin < 3 || isempty(orientations)
    orientations = repmat({'Down (-Z)'}, nWP, 1);
end
if nargin < 4 || isempty(toolSpeeds)
    toolSpeeds = 0.20 * ones(nWP,1);
elseif isscalar(toolSpeeds)
    toolSpeeds = toolSpeeds * ones(nWP,1);
end

q0 = homeConfiguration(robot);
q_wps = zeros(7, nWP+1);
q_wps(:,1) = q0;

ik = inverseKinematics('RigidBodyTree', robot);
ik.SolverParameters.AllowRandomRestart = false;
ik.SolverParameters.MaxNumIteration   = 1500;

%% IK for every waypoint
for w = 1:nWP
    tx=waypoints_xyz(w,1); ty=waypoints_xyz(w,2); tz=waypoints_xyz(w,3);
    ori_ = oriString2tform(orientations{w});
    T_goal = trvec2tform([tx ty tz]) * ori_;
    q_s = wrapToPi(ik(ee, T_goal, [1 1 1 1 1 1], q_wps(:,w)));
    Tc  = getTransform(robot, q_s, ee);
    em  = norm(Tc(1:3,4) - [tx;ty;tz]) * 1000;
    if em > 5
        error('generateMultiWaypoint: WP%d unreachable (%.1f mm)', w, em);
    end
    q_wps(:,w+1) = q_s;
end

%% Segment durations (per-WP speed)
seg_T = zeros(1,nWP);
for w = 1:nWP
    Ts = getTransform(robot, q_wps(:,w),   ee);
    Te = getTransform(robot, q_wps(:,w+1), ee);
    d  = norm(tform2trvec(Ts) - tform2trvec(Te));
    seg_T(w) = max(round(d/toolSpeeds(w), 1), 1.5);
end

%% Junction velocities
jv = zeros(7, nWP+1);
for w = 2:nWP
    di = (q_wps(:,w)   - q_wps(:,w-1)) / seg_T(w-1);
    do_= (q_wps(:,w+1) - q_wps(:,w))   / seg_T(w);
    vj = (seg_T(w).*di + seg_T(w-1).*do_) / (seg_T(w-1)+seg_T(w));
    for j=1:7
        if sign(di(j)) ~= sign(do_(j)); vj(j)=0; end
    end
    jv(:,w) = vj;
end

%% Build trajectory
all_t=[]; all_p=[]; all_v=[]; all_a=[]; t_off=0;
for w = 1:nWP
    qs=q_wps(:,w); qe=q_wps(:,w+1);
    qds=jv(:,w); qde=jv(:,w+1); T_=seg_T(w);
    ts=(0:dt:T_)'; Ns=length(ts);
    p_=zeros(7,Ns); v_=zeros(7,Ns); a_=zeros(7,Ns);
    for j=1:7
        a0=qs(j); a1=qds(j);
        a2=(3*(qe(j)-qs(j))/T_^2)-(2*qds(j)/T_)-(qde(j)/T_);
        a3=(-2*(qe(j)-qs(j))/T_^3)+((qds(j)+qde(j))/T_^2);
        t=ts';
        p_(j,:)=a0+a1.*t+a2.*t.^2+a3.*t.^3;
        v_(j,:)=a1+2*a2.*t+3*a3.*t.^2;
        a_(j,:)=2*a2+6*a3.*t;
    end
    if w<nWP; p_=p_(:,1:end-1); v_=v_(:,1:end-1); a_=a_(:,1:end-1); ts=ts(1:end-1); end
    all_t=[all_t; ts+t_off]; all_p=[all_p,p_]; all_v=[all_v,v_]; all_a=[all_a,a_];
    t_off = t_off + ts(end) + dt;
end

result.q_ref_traj   = [all_t, all_p'];
result.qd_ref_traj  = [all_t, all_v'];
result.tau_ref_traj = [all_t, all_a'];
result.T_total      = all_t(end);
result.seg_T        = seg_T;
result.ik_err_mm    = 0;   % checked per-WP above
result.Kp = [100;100;80;60;40;40;20] * 3.0;
result.Kd = [20;20;16;12;8;8;4]     * 3.0;
end

function T = oriString2tform(s)
switch strtrim(s)
    case 'Up (+Z)';          T = eye(4);
    case 'Horizontal (+X)';  T = axang2tform([0 0 1 pi/2]);
    case 'Tilted 45 deg';    T = axang2tform([0 1 0 pi*3/4]);
    otherwise;               T = axang2tform([0 1 0 pi]);  % Down (-Z)
end
end

File: generate_multi_waypoint.py
"""
generate_multi_waypoint.py
Multi-waypoint cubic Hermite trajectory with junction velocity blending.
Python equivalent of generateMultiWaypoint.m
"""

import numpy as np
from robot_model import KinovaGen3


def generate_multi_waypoint(robot: KinovaGen3,
                             waypoints: np.ndarray,
                             orientations: list,
                             speeds: np.ndarray) -> dict:
    """
    Build a multi-waypoint cubic Hermite trajectory.

    Parameters
    ----------
    robot        : KinovaGen3 instance
    waypoints    : (N x 3) array of [x, y, z] targets in metres
    orientations : list of N strings ('Down','Up','Horiz','Tilt45')
    speeds       : (N,) array of tool speeds in m/s per segment

    Returns
    -------
    dict with keys:
        q_ref_traj    : (N x 8) [time, q1..q7]
        qd_ref_traj   : (N x 8)
        tau_ref_traj  : (N x 8) accelerations rad/s²
        T_total       : float
        seg_T         : list of segment durations
        ik_err_mm     : float (max across waypoints)
    """
    dt     = 0.001
    n_wps  = len(waypoints)

    # ── IK for each waypoint ──────────────────────────────────────────────────
    print(f'  [MW] Solving IK for {n_wps} waypoints...')
    q_wps    = np.zeros((7, n_wps + 1))   # +1 for home start
    q_wps[:, 0] = robot.home_config()
    max_ik_err = 0.0

    q_prev = robot.home_config()
    for w in range(n_wps):
        ori = orientations[w] if orientations[w] else 'Down'
        T_target = robot.target_transform(waypoints[w], ori.lower())
        q_sol, ok, err = robot.ikine(T_target, q_prev)
        if not ok:
            raise ValueError(f'IK failed at waypoint {w+1}: err={err:.1f}mm')
        q_wps[:, w+1] = q_sol
        max_ik_err = max(max_ik_err, err)
        q_prev = q_sol
        print(f'    WP{w+1}: IK={err:.3f}mm  q[1]={np.rad2deg(q_sol[1]):.1f}°')

    # ── Segment timings ───────────────────────────────────────────────────────
    seg_T = []
    for w in range(n_wps):
        qs = q_wps[:, w]
        qe = q_wps[:, w+1]
        dist = np.linalg.norm(qs - qe)
        spd  = speeds[w] if w < len(speeds) else 0.20
        seg_T.append(max(round(dist / spd, 1), 1.0))

    print(f'  [MW] Segment times: {[round(t,1) for t in seg_T]} s')

    # ── Junction velocity blending ────────────────────────────────────────────
    # Catmull-Rom style: average of adjacent segment velocities
    n_segs = n_wps
    jv = np.zeros((7, n_segs + 1))  # junction velocities (zero at start/end)

    for w in range(1, n_segs):
        di = (q_wps[:, w]   - q_wps[:, w-1]) / seg_T[w-1]
        do = (q_wps[:, w+1] - q_wps[:, w])   / seg_T[w]
        vj = (seg_T[w] * di + seg_T[w-1] * do) / (seg_T[w-1] + seg_T[w])
        # Zero velocity at direction reversals
        for j in range(7):
            if np.sign(di[j]) != np.sign(do[j]):
                vj[j] = 0.0
        jv[:, w] = vj

    # ── Build cubic Hermite segments ──────────────────────────────────────────
    all_t   = []
    all_pos = []
    all_vel = []
    all_acc = []
    t_offset = 0.0

    for w in range(n_segs):
        qs  = q_wps[:, w]
        qe  = q_wps[:, w+1]
        qds = jv[:, w]
        qde = jv[:, w+1]
        T_  = seg_T[w]

        ts = np.arange(0, T_ + dt, dt)
        Ns = len(ts)
        p_ = np.zeros((7, Ns))
        v_ = np.zeros((7, Ns))
        a_ = np.zeros((7, Ns))

        for j in range(7):
            a0 = qs[j];  a1 = qds[j]
            a2 = (3*(qe[j]-qs[j])/T_**2) - (2*qds[j]/T_) - (qde[j]/T_)
            a3 = (-2*(qe[j]-qs[j])/T_**3) + ((qds[j]+qde[j])/T_**2)
            p_[j] = a0 + a1*ts + a2*ts**2 + a3*ts**3
            v_[j] = a1 + 2*a2*ts + 3*a3*ts**2
            a_[j] = 2*a2 + 6*a3*ts

        # Remove duplicate junction point (except last segment)
        if w < n_segs - 1:
            p_ = p_[:, :-1]; v_ = v_[:, :-1]; a_ = a_[:, :-1]
            ts = ts[:-1]

        all_t.append(ts + t_offset)
        all_pos.append(p_)
        all_vel.append(v_)
        all_acc.append(a_)
        t_offset += ts[-1] + dt

    # ── Concatenate ───────────────────────────────────────────────────────────
    t_full   = np.concatenate(all_t)
    pos_full = np.concatenate(all_pos, axis=1)
    vel_full = np.concatenate(all_vel, axis=1)
    acc_full = np.concatenate(all_acc, axis=1)
    T_total  = t_full[-1]

    time_col = t_full.reshape(-1, 1)
    Kp_base = np.array([100,100,80,60,40,40,20], dtype=float)
    Kd_base = np.array([ 20, 20,16,12, 8, 8, 4], dtype=float)
    return {
        'q_ref_traj':   np.hstack([time_col, pos_full.T]),
        'qd_ref_traj':  np.hstack([time_col, vel_full.T]),
        'tau_ref_traj': np.hstack([time_col, acc_full.T]),
        'T_total':      T_total,
        'seg_T':        seg_T,
        'ik_err_mm':    max_ik_err,
        'Kp':           Kp_base * 3.0,
        'Kd':           Kd_base * 3.0,
    }


# ── Quick test ────────────────────────────────────────────────────────────────
if __name__ == '__main__':
    print("Loading robot...")
    robot = KinovaGen3()

    waypoints = np.array([
        [0.5,  0.0, 0.3],
        [0.4,  0.3, 0.4],
        [0.3, -0.2, 0.35],
    ])
    orientations = ['Down', 'Down', 'Down']
    speeds       = np.array([0.20, 0.15, 0.20])

    print("\nGenerating 3-waypoint trajectory...")
    res = generate_multi_waypoint(robot, waypoints, orientations, speeds)

    print(f"\nResults:")
    print(f"  T_total:      {res['T_total']:.2f} s")
    print(f"  Seg times:    {[round(t,1) for t in res['seg_T']]} s")
    print(f"  Max IK error: {res['ik_err_mm']:.3f} mm")
    print(f"  N steps:      {res['q_ref_traj'].shape[0]}")
    print(f"  Shape:        {res['q_ref_traj'].shape}")
    print(f"  q_start:      {np.round(np.rad2deg(res['q_ref_traj'][0,1:]),2)} deg")
    print(f"  q_end:        {np.round(np.rad2deg(res['q_ref_traj'][-1,1:]),2)} deg")
    print(f"  vel_start:    {np.round(res['qd_ref_traj'][0,1:],4)}")
    print(f"  vel_end:      {np.round(res['qd_ref_traj'][-1,1:],4)}")

    print("\ngenerate_multi_waypoint.py — OK")